### ISSUE TITLE
A less black-boxy `vim.secure`

### ISSUE THREAD

**Issue by HiPhish** [neovim/neovim#27686]
### Problem

Currently the `vim.secure` module is very restrictive. The only way to check a file is to try to read it, and there is no automated way to allow a file. This is a problem in automated testing where I want to create and destroy a trust DB on the fly.

### Expected behavior

I would like a way to perform these actions:

- Check whether a file is trusted without performing any side effects like reading
- Allow a file without asking for confirmation

I think the first point is not controversial at all. However, the second point might be. In the discussions over at #27627 this was considered a vulnerability, but I disagree. Case in point, in [a test](https://github.com/HiPhish/neotest-busted/blob/f4b3ad0d1af90961b6bd903c94e5a6922c1e8ece/test/integration/config-detection.lua#L15) of mine I manually compute the hashes and create a trust DB file, so if I can circumvent `vim.secure` then so can any malicious plugin.

My understanding is that `vim.secure.read` exists to prevent executing *unexpected* code from *unknown* files like in the `exrc` option. For example, when I download a codebase and open the code in my editor to inspect it, I don't want to start executing code from that repo without confirmation. However, when I install a plugin I have made a deliberate choice. At this point the plugin can just append a line directly to the trust DB, it does not have to go through `vim.secure`. Not to mention that the plugin can do damage directly, whitelisting a foreign file which the user may or may not ever open is a pointlessly roundabout attack.

**Comment by justinmk** [comment]
> elaborate what the purpose of `:trust` is then?

Same as Windows UAC :) It explicitly asks the user if they want to run a script. Automating that defeats its purpose.

> I am trying to understand why I cannot allow a file through scripting

scripts can call `:trust` , which accepts a `[file]` arg.

**Comment by HiPhish** [comment]
> scripts can call `:trust` , which accepts a `[file]` arg.

I tried that, I get `E475: Invalid argument: src/arithmetic.fnl` (or whatever the file is), otherwise I would have used that. Is this a bug?

To provide some context, I recently wrote [HiPhish/neotest-busted](https://github.com/HiPhish/neotest-busted), which loads the `.busted` file to read in the user's settings. Since `.busted` can contain arbitrary Lua code I use `trust.read` to first ask the user for confirmation. Of course I also need to test whether the feature actually works, so in one of my integration tests I generate a trust DB which has some files allowed and some not, and then use that trust DB during testing.

**Comment by justinmk** [comment]
> I tried that, I get `E475: Invalid argument: src/arithmetic.fnl` (or whatever the file is), otherwise I would have used that. Is this a bug?

what exact code did you try? we have tests: https://github.com/neovim/neovim/blob/5d4e1693cb415e8b76749e833e28f00f14630b87/test/functional/ex_cmds/trust_spec.lua#L59

It's possible that we require you to be viewing the file before trusting it.

**Comment by HiPhish** [comment]
> what exact code did you try?

I executed `:trust src/arithmetic.fnl` or `:trust %`, both throw the same error, regardless of whether the file is in the current buffer or not. Executing `:trust` without any argument works fine.

Neovim version 0.9.5

**Comment by justinmk** [comment]
yes `:help :trust` makes it clear, the `[file]` argument is only relevant for ++deny/++remove. So `:edit` the file first. Scripts can do that :)

**Comment by HiPhish** [comment]
> Scripts can do that :)

Yes, but only by mutating the editor state, which means I can only use that approach in end-to-end tests, not unit tests. I think I'll just stick to writing to the trust DB file directly instead.
